WordPress/gutenberg
# Storybook: Support proper extraction of TypeScript prop types

## Description

This PR clears the path for proper extraction of component prop types to use in our Storybook and documentation. Given our ongoing investment in TypeScript migrations, it would be great to leverage this as much as possible.

I tweaked two components here (`Heading` and `Divider`) to demonstrate what it could look like. A big benefit is that the prop type documentation **and** all Controls can be generated automatically for the most part, with inline descriptions ✨   In some cases we may want to override or hide specific controls, but other than that we get everything for free.

### Details

After a lot of experimentation, I figured out some things that were preventing Storybook's `react-docgen-typescript` from properly extracting props information from our TypeScript/JSDocs. Here's what's required for proper type extraction:

- The component has to be written in actual TypeScript, not just typed with a JSDoc comment 😅 
- The `WordPressComponentProps` type alias needs to be imported from a proper `.ts` source file. The problem was that our components were importing it via `ui/context/index.js`, which is not a `.ts` file.
- The `contextConnect()` HOC needs to be in TypeScript.
- The connected component needs to be exported individually as a named export, not just as `default`.

### What this means for TypeScript migration and story authoring

As more components are converted to TypeScript and their stories are improved, we can expect that **most consumer devs will want to interact with Storybook primarily through the Docs view**, rather than the Canvas. Especially with code samples (92354c96a3f1e8a698284a981bf2ffb05fbb3d62) visible for each story, it's much easier to understand how to use the components. And for the default story, the Controls are available right in the Docs view. Canvas view will only be necessary for advanced debugging.

Eventually, I imagine we'll want to make the [Docs mode version](https://storybook.js.org/docs/react/writing-docs/build-documentation) of our Storybook to be our canonical documentation, instead of the limited [Component Reference](https://developer.wordpress.org/block-editor/reference-guides/components/) currently on the dotorg site.

So whenever possible, it'll be beneficial for us to write TypeScript and Storybook stories in a way that makes sense for the Docs view. For example:

- Are there any weird props that show up in Docs view that you want to hide? Consider whether it should be `Omit`-ed at the TypeScript level. (Example: #38844)
- Are some stories showing unhelpful code snippets in Docs view? See if you can tweak the story to show an understandable code sample.

## Testing Instructions

1. `npm run storybook:dev`
2. See stories for `Heading` and `Divider`. Compare with the ones on https://wordpress.github.io/gutenberg/
3. If you're interested, try `npx start-storybook -c ./storybook -p 50240 --docs` to see how it would look like in Docs mode.

## Screenshots 

https://user-images.githubusercontent.com/555336/154171498-07d10d0f-3f43-4509-b3cb-346205619cfa.mp4

## Types of changes

- Minor, non-consumer-facing changes (Changing files to `.ts`, re-exporting as named components)
- Storybook changes

## Checklist:
